### PR TITLE
disttask, ddl: dispatch step-two tasks to proper instances

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: ref #46258

Problem Summary:

Create a cluster with two TiDB instances.

```sql
mysql> use test
Database changed
mysql> set global tidb_enable_dist_task=1;
Query OK, 0 rows affected (0.02 sec)

mysql> create table t(a int);
Query OK, 0 rows affected (0.13 sec)

mysql> insert into t values(1);
Query OK, 1 row affected (0.01 sec)

mysql> alter table t add index idx(a);
ERROR 8247 (HY000): Ingest failed: ingest engine not found
```

If a TiDB has never run read-index stage(step one), no ingest engine is created. When it tries to run ingest-stage(step two), "ingest engine not found" error is reported.

Add index dispatcher should only dispatch step-two tasks to those instances that have run step-one.

### What is changed and how it works?

- Consider previous instance IDs in `GetEligibleInstances()` of add index dispatcher.
- Other tiny code refactoring.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
  
  It is not easy to mock multi-instance environment in unit tests for functional tests.
  ```sql
  mysql> set global tidb_enable_dist_task=1;
  Query OK, 0 rows affected (0.02 sec)

  mysql> create table t(a int);
  Query OK, 0 rows affected (0.11 sec)

  mysql> insert into t values(1);
  Query OK, 1 row affected (0.02 sec)

  mysql> alter table t add index idx(a);
  Query OK, 0 rows affected (4.80 sec)

  mysql> admin check table t;
  Query OK, 0 rows affected (0.00 sec)
  ```
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by ywqzzy** [comment]
Maybe we can refine the `processNormalFlow` interface to make it return ([][]byte, flag bool, error)
If in next stage we want to dispatch the same subtask to nodes that run previous stage.
Just return one subtask and set the bool return value to true.
If the flag == true Then when we dispatch subtasks, just send the subtask to nodes that run previous stage.
Else we use the previous round robin way to dispatch subtasks.

File: disttask/framework/storage/task_table.go [hunk 1]
```diff
@@ -510,6 +510,26 @@ func (stm *TaskManager) GetSchedulerIDsByTaskID(taskID int64) ([]string, error)
 	return instanceIDs, nil
 }
 
+// GetSchedulerIDsByTaskIDAndStep gets the scheduler IDs of the given global task ID and step.
+func (stm *TaskManager) GetSchedulerIDsByTaskIDAndStep(taskID int64, step int64) ([]string, error) {
```
Comment (ywqzzy) [inline disttask/framework/storage/task_table.go:514]: Add ut?

**Review by ywqzzy — APPROVED** [review]
Rest LGTM

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Comment by tangenta** [comment]
Wait for https://github.com/pingcap/tidb/pull/46671

File: util/disttask/idservice.go [hunk 2]
```diff
@@ -32,13 +32,18 @@ func GenerateExecID(ip string, port uint) string {
...
+func FindServerInfo(serverInfos []*infosync.ServerInfo, schedulerID string) int {
+	for i, serverInfo := range serverInfos {
 		serverID := GenerateExecID(serverInfo.IP, serverInfo.Port)
 		if serverID == schedulerID {
-			return true
+			return i
 		}
 	}
-	return false
+	return -1
```
Comment (Benjamin2037) [inline util/disttask/idservice.go:46]: Should we use a InvalidServerID const to mean -1?

**Pull request merged**